pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -57,20 +71,220 @@ type defaultStatefulSetControl struct {
+
+	// get the current, and update revisions
+	currentRevision, updateRevision, err := ssc.getStatefulSetRevisions(set, revisions)
+	if err != nil {
+		return err
+	}
+
+	// perform the main update function and get the status
+	status, err := ssc.updateStatefulSet(set, currentRevision, updateRevision, pods)
+	if err != nil {
+		return err
+	}
+
+	// update the set's status
+	err = ssc.updateStatefulSetStatus(set, status)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("StatefulSet %s/%s pod status replicas=%d ready=%d current=%d updated=%d",
+		set.Namespace,
+		set.Name,
+		status.Replicas,
+		status.ReadyReplicas,
+		status.CurrentReplicas,
+		status.UpdatedReplicas)
+
+	glog.V(4).Infof("StatefulSet %s/%s revisions current=%s update=%s",
+		set.Namespace,
+		set.Name,
+		status.CurrentRevision,
+		status.UpdateRevision)
+
+	// maintain the set's revision history limit
+	return ssc.truncateHistory(set, pods, revisions, currentRevision, updateRevision)
+}
+
+func (ssc *defaultStatefulSetControl) ListRevisions(set *apps.StatefulSet) ([]*apps.ControllerRevision, error) {
+	selector, err := metav1.LabelSelectorAsSelector(set.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	return ssc.controllerHistory.ListControllerRevisions(set, selector)
+}
+
+func (ssc *defaultStatefulSetControl) AdoptOrphanRevisions(
+	set *apps.StatefulSet,
+	revisions []*apps.ControllerRevision) error {
+	for i := range revisions {
+		adopted, err := ssc.controllerHistory.AdoptControllerRevision(set, controllerKind, revisions[i])
+		if err != nil {
+			return err
+		}
+		revisions[i] = adopted
+	}
+	return nil
+}
+
+// truncateHistory truncates any non-live ControllerRevisions in revisions from set's history. The UpdateRevision and
+// CurrentRevision in set's Status are considered to be live. Any revisions associated with the Pods in pods are also
+// considered to be live. Non-live revisions are deleted, starting with the revision with the lowest Revision, until
+// only RevisionHistoryLimit revisions remain. If the returned error is nil the operation was successful. This method
+// expects that revisions is sorted when supplied.
+func (ssc *defaultStatefulSetControl) truncateHistory(
+	set *apps.StatefulSet,
+	pods []*v1.Pod,
+	revisions []*apps.ControllerRevision,
+	current *apps.ControllerRevision,
+	update *apps.ControllerRevision) error {
+	history := make([]*apps.ControllerRevision, 0, len(revisions))
+	// mark all live revisions
+	live := map[string]bool{current.Name: true, update.Name: true}
+	for i := range pods {
+		live[getPodRevision(pods[i])] = true
+	}
+	// collect live revisions and historic revisions
+	for i := range revisions {
+		if !live[revisions[i].Name] {
+			history = append(history, revisions[i])
+		}
+	}
+	if len(history) < int(*set.Spec.RevisionHistoryLimit) {
+		return nil
+	}
+	// delete any non-live history to maintain the revision limit.
+	history = history[:int(*set.Spec.RevisionHistoryLimit)]
```

Comment:
origiHistory = [0 1 2 3 4 ]
revisionLimit  = 3
newHistory = origHistory[ :3 ]
newHistory = [0 1 2] which is deleted in the subsequent loop. so the remaining left is 3 and 4. What we really want is 0 and 1 oldest revision deleted and the only remaining should be [2,3 ,4]
so history = history[ : len(history)-revisionLimit]  @kow3ns ?